Add Baseline status to hovercards

rviscomi opened the pull request:
Progress on #427

- [x] sync custom-data dependency with the output of https://github.com/microsoft/vscode-custom-data/pull/111
- [x] add Baseline info in hovercards
- [x] update tests

Hovercard changes:
- adds baseline icon and status
  - replaces the browser version list
  - icon supports light/dark mode and corresponds to the feature status
- omits the experimental warning (covered by limited availability status)

Demo:

https://github.com/user-attachments/assets/8cf066b2-2e09-4c95-9579-51423cab7f12

<details>

<summary>Screenshots...</summary>

### Widely available

Before:
<img width="939" alt="image" src="https://github.com/user-attachments/assets/4ae2a594-b6f0-48b2-b772-529de60cb712" />

After (dark mode):
<img width="918" alt="image" src="https://github.com/user-attachments/assets/f7279602-2250-4a75-aa0d-02db8745549b" />

After (light mode):
<img width="935" alt="image" src="https://github.com/user-attachments/assets/5449a169-35d8-49f4-9d0f-d27eb2c7aa03" />


### Newly available

Before:
<img width="582" alt="image" src="https://github.com/user-attachments/assets/20f1e18b-2e79-4c4e-86ac-8d9025af3ea1" />

After (dark mode):
<img width="561" alt="image" src="https://github.com/user-attachments/assets/67697e03-376c-41a8-bdf6-0d0a3c7eebd6" />

After (light mode):
<img width="561" alt="image" src="https://github.com/user-attachments/assets/6116b1f2-46f1-4bc1-8ab6-39c80706cd46" />

### Limited availability

Before:
<img width="940" alt="image" src="https://github.com/user-attachments/assets/3fcc656e-d52f-4c61-9a2b-fb4b3f11d406" />

After (dark mode):
<img width="918" alt="image" src="https://github.com/user-attachments/assets/a7e8eb5c-fa7b-4293-b328-ba5aaf5d3684" />

After (light mode):
<img width="920" alt="image" src="https://github.com/user-attachments/assets/7f217d70-cdc7-45c9-9529-9d2e5ebddb3d" />

### Experimental

Before:
<img width="932" alt="image" src="https://github.com/user-attachments/assets/8b22dfe7-0123-422f-82e2-0c7f54e9d8ed" />

After (dark mode):
<img width="903" alt="image" src="https://github.com/user-attachments/assets/fdcb8406-6e12-46a7-a43e-66d299a2a996" />

After (light mode):
<img width="903" alt="image" src="https://github.com/user-attachments/assets/966f0bb2-cbf6-4ad3-ad94-bb3deda2f2b2" />

(experimental warning removed, as it's redundant with the limited availability message)

### Obsolete

Before:
<img width="934" alt="image" src="https://github.com/user-attachments/assets/18c8f2d8-c3f3-4a12-b89b-5825f981ee73" />

After (dark mode):
<img width="918" alt="image" src="https://github.com/user-attachments/assets/755f5ee5-05ca-414a-88d1-f6b03a607795" />

After (light mode): 
<img width="923" alt="image" src="https://github.com/user-attachments/assets/b09ae962-c6dc-45cc-bc42-499040b69b0b" />

(obsolete warning kept)

</details>

rviscomi commented:
FYI I was getting a TypeScript error related to my use of the `Intl.ListFormat` function:

```
$ npm run test

> vscode-css-languageservice@6.3.2 test
> npm run compile && npm run mocha


> vscode-css-languageservice@6.3.2 compile
> tsc -p ./src && npm run copy-jsbeautify && npm run lint

src/languageFacts/entry.ts:193:50 - error TS2339: Property 'ListFormat' does not exist on type 'typeof Intl'.

193 const missingBaselineBrowserFormatter = new Intl.ListFormat("en", {
                                                     ~~~~~~~~~~


Found 1 error in src/languageFacts/entry.ts:193
```

I can fix it by upgrading the tsconfig to es2021, but figured that would need to be part of a separate discussion. For now I worked around it by changing it to `(Intl as any)`.

captainbrosset commented on src/languageFacts/entry.ts:
```diff
@@ -141,32 +196,34 @@ function getEntryMarkdownDescription(entry: IEntry2, settings?: HoverSettings):
+	}
+
+	return missingBaselineBrowserFormatter.format(Object.values(Array.from(missingBrowsers.entries()).reduce((browsers: Record<string, string>, [browserId, browser]) => {
+		if (browser.name in browsers || browserId == 'edge') {
```
The core browser set is not likely to change much, but if it does, this hard cording (as well as the `browserNames` const are a bit unfortunate as they'll need to be changed if the core browser set changes.
Can you think of a utility package/function which we could maintain on the web-features repo instead, which would avoid consumers like this one to have to do this?

rviscomi replied:
Hm actually the `computeBaseline` function does already expose browser names and platforms, eg:

```
Browser {
    id: 'chrome',
    data: {
      accepts_flags: true,
      accepts_webextensions: true,
      name: 'Chrome',
      pref_url: 'chrome://flags',
      preview_name: 'Canary',
      releases: [Object],
      type: 'desktop'
    },
    ...
}
```

I'll look into moving the "missing browser" functionality over to [vscode-custom-data#111](https://github.com/microsoft/vscode-custom-data/pull/111/files#diff-df39d64e75bdd56b21124ae2fcdc9778ae695292a82c50630bd1ebfdbc9e9f2aR374) where it already uses `computeBaseline`.

rviscomi replied:
Given @aeschli's [comment](https://github.com/microsoft/vscode-custom-data/pull/111#issuecomment-2740307413) about the need for a well-defined schema, I think this is actually ok. If the core browser set changes, this implementation will need to be updated in a few places anyway.

aeschli commented:
````
      "browsers": [
        "E12",
        "S10",
        "C50",
        "IE10",
        "O37"
      ],
      "baseline": {
        "status": "high",
        "support": {
          "chrome": "50",
          "chrome_android": "50",
          "edge": "12",
          "firefox": "28",
          "firefox_android": "28",
          "safari": "10",
          "safari_ios": "10"
        },
        "baseline_low_date": "2015-09-30",
        "baseline_high_date": "2018-03-30"
      },
 ```


Now the browser information is duplicated, or have conflicting information? Can't we use the browsers we already have? 
It's ok to add new browsers (just need to spec it)
If it's incomplete or has old data, we can set the data based on the baseline call.

rviscomi commented:
@aeschli sure I've removed the `baseline.support` object and added the mobile versions of the Baseline core browser set to the existing `browsers` array

rviscomi commented:
Found an issue with the [`accent-color`](https://developer.mozilla.org/en-US/docs/Web/CSS/accent-color) property, which is not Baseline due to an implementation issue in Safari. In BCD, this is indicated by the [`partial_implementation`](https://github.com/mdn/browser-compat-data/blob/c6992676d95f20b9d9be3a06726cd609e8a2800d/css/properties/accent-color.json#L27-L31) flag. `compute-baseline` takes this into consideration and omits Safari from the list of supported browsers. But Safari still technically has a supported version in BCD and ends up in the list.

Actual:

<img width="389" alt="image" src="https://github.com/user-attachments/assets/0220e4c0-49d3-45f0-8525-fb78e208a0fc" />

Expected:

<img width="659" alt="image" src="https://github.com/user-attachments/assets/fbe5bae9-3f86-485b-be21-1b38f0147948" />

I fixed this by removing support strings from the `browsers` array for any core browsers missing from Baseline's support list.

aeschli commented:
Thanks for the PR. I think we need to improve the look of the hover.
- I don't find the baseline icon intuitive. I would would go with the ⚠️ and a :heavy_check_mark: 
- I would not put the name first

My suggestion:
- description
- compatibility warning (Property is experimental. Property is obsolete, Property is nonstandard)
- baseline
- syntax

aeschli commented:
The mockups are good. I still don't like the extra wide baseline icons. It's a bit too large for the text size, also I find the form not very intuitive. Is the shape important to you? Can we not just add a checkmark?

aeschli commented:
BTW I adopted the custom data, but didn't make any changes to the hover. Leaving this to you.
https://github.com/microsoft/vscode-css-languageservice/pull/431/files

rviscomi commented:
Here's an updated mockup of a 10px tall icon instead of the 14px version in light mode:

<img width="725" alt="image" src="https://github.com/user-attachments/assets/97d88548-073d-427c-854a-576ec105237e" />

If that looks good to you, I'll push a commit updating the rest of the icons.

captainbrosset commented:
Regarding the use of icons, thank you @aeschli for your feedback on the fact that they're not intuitive. @atopal, let's add this to other feedback we might have received and reflect on whether improvements are needed on the long run.

For now, though (i.e., for this PR), I strongly suggest keeping the Baseline icons as they are defined now, and not deviate from them.
Baseline is a concept that was introduced a couple of years ago and is getting adopted by more and more sites, dashboards, and tools. We pay a lot of attention as to how Baseline is being adopted, and sometimes even work hand in hand with consumers to make sure that Baseline is communicated correctly.
The reason is that Baseline is an attempt at summarizing the many details of a very fragmented and nuanced web platform, into something that's as simple as 3 statuses only. This is essentially impossible to do, but we believe that doing it in the way we're doing it now still has more benefits than drawbacks. If, and only if, Baseline is communicated correctly.

Changing the icons to a checkmark and a warning icon would simplify what Baseline is even more, to the point that it becomes dangerous.
Indeed, Baseline says "if this is widely available, then you don't really need to worry about it, or test much". That's basically all it says.
Baseline **does not** say "if this is not widely available, then don't use it". If a feature is not widely available, then it's basically up to the developer to make  conscious decision as to whether they want to use that feature. They may very well use it, even if the feature has limited availability, if they've tested it for their specific use case. Maybe they're using a polyfill in other browsers. Maybe it's part of a progressively enhanced feature set. Who knows.

So, Baseline messages/icons should never appear to developers as something that may steer them away from using a feature.

aeschli commented:
The smaller baseline icon looks better. Lets go with that!

aeschli commented:
@rviscomi Thanks Rick, nice work, it looks great!

rviscomi commented:
Thanks for all your help and can't wait to see it live! 🥳

FluorescentHallucinogen commented on src/languageFacts/entry.ts:
```diff
@@ -40,6 +56,39 @@ function getEntryStatus(status: EntryStatus) {
+		return status;
+	}
+
+	const baselineYear = baseline.baseline_low_date?.split('-')[0];
```
@rviscomi If the status is `low`, then `baseline_low_date` is used, but if the status is `high`, then again `baseline_low_date` is used, not `baseline_high_date`. It turns out that `baseline_high_date` is not used at all. Is this not a bug? Is this by design?

rviscomi replied:
Hi, yes this is by design and is consistent with other Baseline resources.  For example, the [`dialog`](https://github.com/web-platform-dx/web-features/blob/2898cf81bb72673ecf32d04089d8fa312e6287ad/features/dialog.yml.dist#L1-L7) element is widely available since 2024, and Baseline since 2022:

```
status:
  baseline: high
  baseline_low_date: 2022-03-14
  baseline_high_date: 2024-09-14
```

Resources like [MDN](https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/dialog) and the [Web Platform Dashboard](https://webstatus.dev/features/dialog) show the newly available date:

<img width="793" alt="image" src="https://github.com/user-attachments/assets/f49cf2d9-6113-4c0f-9875-24a589f85ab7" />

<img width="899" alt="image" src="https://github.com/user-attachments/assets/75cf867a-8743-4d3f-ab0b-2c1e1a74c7f3" />

FluorescentHallucinogen commented on src/languageFacts/entry.ts:
```diff
@@ -141,31 +196,39 @@ function getEntryMarkdownDescription(entry: IEntry2, settings?: HoverSettings):
+// TODO: Remove "as any" when tsconfig supports es2021+
+const missingBaselineBrowserFormatter = new (Intl as any).ListFormat("en", {
+	style: "long",
+	type: "disjunction",
```
@rviscomi Why did you use disjunction instead of conjunction here? 🤔

rviscomi replied:
"or" is more correct. Here's Gemini to help articulate exactly why that is 😄 

> When listing items that share a common negative characteristic, using "or" is generally more grammatically precise and less ambiguous.
>
>So, the more correct phrasing is:
>**"Not fully implemented in Chrome, Edge, or Safari"**
>
>Here's why:
>
>1. Meaning with "or": This sentence structure implies that the negation ("Not fully implemented") applies to each browser individually:
>    - It is not fully implemented in Chrome,
>    - AND it is not fully implemented in Edge,
>    - AND it is not fully implemented in Safari. This accurately conveys that none of the listed browsers have full implementation. In logic, this corresponds to `¬A ∧ ¬B ∧ ¬C`.
>2. Potential ambiguity with "and": The sentence "Not fully implemented in Chrome, Edge, and Safari" could be interpreted to mean: "It is not the case that the feature is fully implemented in all three (Chrome, Edge, and Safari) simultaneously."
>    - This statement would be true even if the feature was implemented in Chrome and Edge but not in Safari (i.e., at least one of them doesn't have it). In logic, this corresponds to `¬(A ∧ B ∧ C)`, which is equivalent to `¬A ∨ ¬B ∨ ¬C` (at least one is not implemented).
>    - While readers might often infer the intended meaning (that none of them have it), the "and" phrasing is technically ambiguous and not as precise for indicating that each listed item lacks the feature.
>For clear technical documentation, where precision is important, "or" is the preferred conjunction in this negative context to ensure the meaning is that each of the browsers listed does not support the feature.

FluorescentHallucinogen commented on src/languageFacts/entry.ts:
```diff
@@ -4,33 +4,49 @@
+	}
 };
 
+const shortCompatPattern = /(E|FFA|FF|SM|S|CA|C|IE|O)([\d|\.]+)?/;
```
@rviscomi Why are we looking for the `|` character in the `[\d|\.]` part? 🤔

I've analyzed the `browsers.css-data.json` file from the `@vscode/web-custom-data` package and haven't found a single property that has this character in the `browsers` or  `values.browsers` field.

But I've found 5 properties that have values in the `browsers` field, that don't match current pattern:

`overflow-anchor`: `'Spreview'`
`-moz-background-clip`: `'FF1-3.6'`
`-ms-filter`: `'IE8-9'`
`-ms-transform`: `'IE9-9'`
`-ms-transform-origin`: `'IE9-9'`

Actually, we are not interested in the version value (the second part of the regexp), only the browser name (the first part of the regexp).

Maybe we can do with a simpler regexp or without using a regexp at all?

aeschli replied:
You can remove the `|`...
